File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {  
    public int numIslands(char[][] grid) {
        int count = 0;
        int row = grid.length, col = grid[0].length;
        for(int i=0; i<row; i++){
            for(int j=0; j<col; j++){
                if(grid[i][j] == '1'){
                    count++;
                    dfs(grid,i,j);
                }
            }
        }
        return count;
    }

    public void dfs(char[][] grid, int i, int j){
        if(i<0 || j<0 || i >= grid.length || j >= grid[0].length || grid[i][j] == '0') return;

        grid[i][j] = '0';

        dfs(grid, i+1, j);
        dfs(grid, i-1, j);
        dfs(grid, i, j+1);
        dfs(grid, i, j-1);
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        island_count = 0
        row = len(grid)
        col = len(grid[0])

        if not grid:
            return 0
        
        for r in range(row):
            for c in range(col):
                if grid[r][c] == "1":
                    self.dfs(grid, r, c)
                    island_count += 1
        return island_count

    def dfs(self, grid, r, c) -> None:
        if (r < 0 or c < 0 or r >= len(grid) or c >= len(grid[0]) or grid[r][c] != "1"):
            return

        grid[r][c] = "0"

        self.dfs(grid, r-1, c)
        self.dfs(grid, r+1, c)
        self.dfs(grid, r, c-1)
        self.dfs(grid, r, c+1) 

        